vhost/src/vhost_user/gpu_backend_req.rs

```diff
@@ -1,17 +1,59 @@
 // Copyright (C) 2024 Red Hat, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
+use std::io;
+use std::os::fd::RawFd;
 use std::os::unix::net::UnixStream;
 use std::sync::{Arc, Mutex, MutexGuard};
 
+use vm_memory::ByteValued;
+
+use crate::vhost_user;
 use crate::vhost_user::connection::Endpoint;
 use crate::vhost_user::gpu_message::*;
+use crate::vhost_user::message::VhostUserMsgValidator;
+use crate::vhost_user::Error;
 
 struct BackendInternal {
+    sock: Endpoint<VhostUserGpuMsgHeader<GpuBackendReq>>,
     // whether the endpoint has encountered any failure
     error: Option<i32>,
 }
 
+impl BackendInternal {
+    fn check_state(&self) -> vhost_user::Result<u64> {
+        match self.error {
+            Some(e) => Err(Error::SocketBroken(io::Error::from_raw_os_error(e))),
+            None => Ok(0),
+        }
+    }
+
+    fn send_header<V: ByteValued + Sized + Default + VhostUserMsgValidator>(
+        &mut self,
+        request: GpuBackendReq,
+        fds: Option<&[RawFd]>,
+    ) -> vhost_user::Result<V> {
+        self.check_state()?;
+
+        let hdr = VhostUserGpuMsgHeader::new(request, 0, 0);
+        self.sock.send_header(&hdr, fds)?;
+
+        self.wait_for_ack(&hdr)
+    }
+
+    fn wait_for_ack<V: ByteValued + Sized + Default + VhostUserMsgValidator>(
+        &mut self,
+        hdr: &VhostUserGpuMsgHeader<GpuBackendReq>,
+    ) -> vhost_user::Result<V> {
+        self.check_state()?;
```

Comment:
Should we check if VHOST_USER_PROTOCOL_F_REPLY_ACK is negotiated?

Or the ack is always expected?

Reply:
The ack is always expected for some messages and never expected for other messages. There is no `VHOST_USER_PROTOCOL_F_REPLY_ACK` flag for this protocol. I'll make sure to add a comment in the source code.